From e6de1a09a2f6a6825341e8463866553b77848ed6
From: "Steven J. Hill" <sjhill@mips.com>
Date: Thu, 12 Jul 2012 17:21:31 +0000
Subject: [PATCH] MIPS: uasm: Add INS and EXT instructions.

These are MIPS32R2 instructions for merging and extracting bit fields
from one GPR into another.

Signed-off-by: Steven J. Hill <sjhill@mips.com>

---
 arch/mips/mm/tlbex.c    |    2 ++
 include/asm-mips/uasm.h |   19 ++++++++++++++++++-
 2 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/include/asm-mips/uasm.h b/include/asm-mips/uasm.h
--- a/include/asm-mips/uasm.h
+++ b/include/asm-mips/uasm.h
@@ -63,7 +64,8 @@ enum opcode {
 	insn_beql, insn_bgez, insn_bgezl, insn_bltz, insn_bltzl,
 	insn_bne, insn_daddu, insn_daddiu, insn_dmfc0, insn_dmtc0,
 	insn_dsll, insn_dsll32, insn_dsra, insn_dsrl, insn_dsrl32,
-	insn_dsubu, insn_eret, insn_j, insn_jal, insn_jr, insn_ld,
+	insn_dsubu, insn_eret, insn_ext, insn_ins,
+	insn_j, insn_jal, insn_jr, insn_ld,
 	insn_ll, insn_lld, insn_lui, insn_lw, insn_mfc0, insn_mtc0,
 	insn_or, insn_ori, insn_rfe, insn_sc, insn_scd, insn_sd, insn_sll,
 	insn_sra, insn_srl, insn_subu, insn_sw, insn_tlbp, insn_tlbwi,
@@ -343,6 +347,20 @@ Ip_u2u1msbu3(op)					\
 		build_insn(buf, insn##op, b, a, c);		\
 	}
 
+#define I_u2u1msbu3(op)				\
+ 	static inline void __cpuinit i##op(u32 **buf, unsigned int a,	\
+ 	 	unsigned int b, unsigned int c, unsigned int d)					\
+	{							\
+		build_insn(buf, insn##op, b, a, c+d-1, c);	\
+	}							\
+
+#define I_u2u1mmsbu3(op)				\
+ 	static inline void __cpuinit i##op(u32 **buf, unsigned int a,	\
+ 	 	unsigned int b, unsigned int c, unsigned int d)					\
+	{							\
+		build_insn(buf, insn##op, b, a, d-1, c);	\
+	}							\
+
 #define I_u1u2(op)						\
 	static inline void __cpuinit i##op(u32 **buf, unsigned int a,	\
 	 	unsigned int b)					\
@@ -396,6 +407,8 @@ I_u2u1u3(_drotr)
 I_u2u1u3(_drotr32);
 I_u3u1u2(_dsubu);
 I_0(_eret);
+I_u2u1mmsbu3(_ext);
+I_u2u1msbu3(_ins);
 I_u1(_j);
 I_u1(_jal);
 I_u1(_jr);
diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -154,6 +154,8 @@ static struct insn insn_table[] __uasminitdata = {
 	{ insn_dsrl32, M(spec_op,0,0,0,0,dsrl32_op), RT | RD | RE },
 	{ insn_dsubu, M(spec_op,0,0,0,0,dsubu_op), RS | RT | RD },
 	{ insn_eret, M(cop0_op,cop_op,0,0,0,eret_op), 0 },
+	{ insn_ins, M(spec3_op,0,0,0,0,ins_op), RS | RT | RD | RE },
+	{ insn_ext, M(spec3_op,0,0,0,0,ext_op), RS | RT | RD | RE },
 	{ insn_j, M(j_op,0,0,0,0,0), JIMM },
 	{ insn_jal, M(jal_op,0,0,0,0,0), JIMM },
 	{ insn_jr, M(spec_op,0,0,0,0,jr_op), RS },
-- 
